DRILL-7970: Add URL Parameters to HTTP Plugin

cgivre opened the pull request:
# [DRILL-7970](https://issues.apache.org/jira/browse/DRILL-7970): Add URL Parameters to HTTP Plugin

## Description
Many APIs require that arguments are passed as part of the URL.  For instance, github's API allows you to get information about an organization's repositories with the following URL:  https://api.github.com/orgs/{org}/repos

This PR adds a capability where filters can be pushed down into the URL from a query so that a user could execute a query:

```sql
SELECT * 
FROM api.github
WHERE org='apache'
``` 
In this query, the value for the `org` parameter would be populated in the URL. 

## Documentation
Updated the `README.md` document.  
To enable URL parameters, a user simply must add curly braces around a parameter which should be passed down.  Note that Drill will throw an exception in the event an API has a URL parameter and one is not supplied in the query.

## Testing
Added two additional unit tests to testing class for the HTTP plugin as well as an additional test for the http client.

cgivre commented:
@jfmario, @jaltekruse, @Daddykong,   
Could one of you take a look at this PR and see what you think?    Thanks!!

luocooong commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpPushDownListener.java:
```diff
@@ -82,9 +85,20 @@ public ScanPushDownListener builderFor(GroupScan groupScan) {
+      HttpUrl url = HttpUrl.parse(groupScan.getHttpConfig().url());
+      if (url != null) {
+        List<String> urlParams = SimpleHttp.getURLParameters(url);
+        assert urlParams != null;
```
Is there a old code block? the assert will throw `AssertionError` when false.

cgivre replied:
Fixed.  I actually changed the `getUrlParameters()` method such that if it encounters any error, it returns an empty list rather than `null`.  My sense is that is better design.

luocooong commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+   * @return A list of URL parameters enclosed by curly braces.
+   */
+  public static List<String> getURLParameters(HttpUrl url) {
+    List<String> parameters = new ArrayList<>();
```
Place the `parameters` after the `Matcher matcher` for better. No need to init the variable if the `decodeURL` is null.

cgivre replied:
Fixed

luocooong commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+    List<String> params = SimpleHttp.getURLParameters(url);
+    String tempUrl = SimpleHttp.decodedURL(url);
+    for (String param : params) {
+      if (filters == null) {
```
Place the code block before the `List<String> params`.
```java
if (! hasURLParameters(url)) { }
if (filters == null) { }
```

cgivre replied:
Fixed

luocooong reviewed (commented):
@cgivre Hi, As the new feature, the code quality look good. I left minor comments for the pull request.

paul-rogers commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+
+      String value = filters.get(param);
+      // If the param is not populated, throw an exception
+      if (Strings.isNullOrEmpty(value)) {
```
A Java `null` presumably means the value is missing. That would be a great place to handle defaults. An empty string might be OK:

`https://acertaincompany.com/projects{PROJ}` where the parameter is either `""` or `"/1234"`.

Also, there is nothing stopping me from doing `../foo?bar={bar}` and a blank value might be OK. Note that this "feature" means that we've got to parse the URL to see if the user included any query params before we add any of ours so we get `?bar=xxx&ours=10` and not `?bar=xxx?ours=10`.

cgivre replied:
Fixed in the context of adding defaults.

paul-rogers commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+          .message("API Query with URL Parameters must be populated. Parameter " + param + " must be included in WHERE clause.")
+          .build(logger);
+      } else {
+        tempUrl = tempUrl.replace("/{" + param + "}", "/" + value);
```
Three comments:

1. The above only replaces the first instance. For `.../{ORG}/projects/{ORG}` there will be two entries, so this might be OK.
2. The `param` is of the from the URL, so cases match. But, if we mapped names to lower case to match the case-insensitive column names, we have to remember to use the original case here.
3. The substitution is unstable. If I have the pathological `.../{foo}/{bar}` and in my query I have `foo='{bar}', then the result will be wrong.
4. It turns out that the regex find method can tell us the offset of the matched pattern. It would be better (and would solve the above issues) if we converted the input URL into a list of prefix/pattern values in which the pattern has the start offset, length, name and optional default. The prefix is the part of the URL between patterns. There may be a prefix-only last segment, or some segments with no prefixes. Examples, noting that we do nothing to police where the parameters are placed in the URL:

```text
{url} -->
  [{prefix="", start=0, length=5, name="url"}]

{method}://foo.com/latest -->
  [prefix="", start=0, length=8, name="method"},
   {prefix="://foo.com/latest", start=-1, length=-1, name=null}]

https://foo.com/api/{op} -->
  [{prefix="https://foo.com/api/", start=xx, length=xx, name="op"}]

https://foo.com/api/?op={op} -->
  [{prefix="https://foo.com/api/?op=", start=xx, length=xx, name="op"}]

https://foo.com/api/{org}/projects/{org}/details -->
  [{prefix="https://foo.com/api/", start=xx, length=xx, name="org"},
   {prefix="/projects/", start=xx, length=xx, name="org"},
   {prefix="/details", start=-1, length=-1, name=null}]
```

We do have tests for all of these?

paul-rogers commented on contrib/storage-http/src/test/java/org/apache/drill/exec/store/http/TestHttpPlugin.java:
```diff
@@ -140,12 +141,17 @@ private static void makeMockConfig() {
     HttpApiConfig mockXmlConfig = new HttpApiConfig("http://localhost:8091/xml", "GET", headers,
       "basic", "user", "pass", null, null, "results", null, "xml", 2,false);
 
+    HttpApiConfig mockGithubWithParam = new HttpApiConfig("http://localhost:8091/orgs/{org}/repos", "GET", headers,
```
Please add tests for the various cases above, along with:

* Duplicate names
* Mixed-case names (param="org", column="ORG", etc.)
* Missing column name
* Column of the wrong type
* Parameter is in the query clause (something that should be tested anyway, url: "...?foo=bar"

cgivre replied:
Added additional unit tests for these cases.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+      decodedURL = URLDecoder.decode(url.toString(), "UTF-8");
+    } catch (UnsupportedEncodingException e) {
+      return null;
+    }
```
Done

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpPushDownListener.java:
```diff
@@ -45,6 +47,7 @@
  * <li>An AND'ed set of such expressions,</li>
  * <li>If the value is one with an unambiguous conversion to
  * a string. (That is, not dates, binary, maps, etc.)</li>
+ * <li>A field from the URL.  For instance in some APIs you have parameters that are part of the path.</li>
```
Fixed.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -57,6 +63,7 @@
  */
 public class SimpleHttp {
   private static final Logger logger = LoggerFactory.getLogger(SimpleHttp.class);
+  private static final Pattern URL_PARAM_REGEX = Pattern.compile("\\{([A-Za-z0-9_]+)\\}");
```
Fixed

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+      return url.toString();
+    }
+
+    List<String> params = SimpleHttp.getURLParameters(url);
```
I thought about that... In most applications I saw, this would likely not occur.  From my observations, when URL parameters are used, it is typically to model some sort of hierarchy.  For instance `somestore.com/orders/{customer_id}/year/month` or something like that.

cgivre replied:
I added a unit test and in this case, if the user has duplicate parameters in the URL they do get mapped:  IE:

```
http://foo.com/{f1}/something/{f1}/other
```

If the user then submits a query like this:

```sql
SELECT ....
FROM ...
WHERE fl='foo'
```

That will generate the following URL:  foo.com/foo/something/foo/other.  With that said, I can't imagine that an API will actually do this, but at least the plugin won't error out.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+          .build(logger);
+      }
+
+      String value = filters.get(param);
```
I converted this to use  a `CaseInsensitiveMap`.   Thus... it is now case insensitive.   Added unit test for the URL mapper.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+      param = matcher.group();
+      param = param.replace("{", "");
+      param = param.replace("}", "");
+      parameters.add(param);
```
I was under the impression that string manipulation functions were less expensive than using regex.  That was why I implemented it this way.  However, if the regex is more efficient, I'll refactor.

cgivre replied:
Fixed.

cgivre commented on contrib/storage-http/README.md:
````diff
@@ -45,6 +45,23 @@ The `connection` property can accept the following options.
+WHERE org = 'apache'
+```
+
+This query would replace the `org`in the URL with the value from the `WHERE` clause, in this case `apache`. 
````
This is a question of how much bad user behavior we want to allow.  I did write unit tests for all these conditions, and it does work.  I also added default values per your suggestion.

cgivre commented:
@paul-rogers Thanks for the review.  I believe I addressed all your comments.  I added the ability to specify a default value in the URL, as well as added a lot of new unit tests.  I fixed a small bug as well with the `errorOn400` config option and added a test for that. 

Some of your questions made me wonder how much "misuse" we should tolerate from a user.   For instance, putting the entire URL as a URL parameter...  This is perhaps foolish to do that, but I didn't think there was a risk here because in order to do that, the user would have to have admin access to the plugin configuration anyway.

paul-rogers commented on contrib/storage-http/README.md:
```diff
@@ -45,6 +45,30 @@ The `connection` property can accept the following options.
+URL:  https://github.com/orgs/{org}/repos
+
+As of Drill 1.20.0, you can simply set the URL in the connection using the curly braces.  If your API includes URL parameters you MUST
+include them in the `WHERE` clause in your query, or Drill will throw an error. 
```
Here we say "MUST", but below we walk it back. Can we improve the flow?

cgivre replied:
Fixed

paul-rogers commented on contrib/storage-http/README.md:
```diff
@@ -45,6 +45,30 @@ The `connection` property can accept the following options.
+#### Limitations on URL Parameters
+* Drill does not support combinations of parameters in queries.  For instance, for the above example, you could not include `WHERE org='apache' OR org='linux'`
+* All URL parameter clauses must be equality only. 
+* 
```
Stray asterisk?

cgivre replied:
Fixed

paul-rogers commented on contrib/storage-http/README.md:
```diff
@@ -45,6 +45,30 @@ The `connection` property can accept the following options.
+{param1}/{param2=default}`.  In this case, the default would be used if and only if there isn't a parameter supplied in the query. 
+
+#### Limitations on URL Parameters
+* Drill does not support combinations of parameters in queries.  For instance, for the above example, you could not include `WHERE org='apache' OR org='linux'`
```
"combinations of parameters" could be misleading. Can I "combine" query parameters with URL parameters? Of course. I thing what you mean is that Drill does not support multiple values for a single parameter: `OR` or `IN` expressions.

cgivre replied:
Ah yes.. I'll update the docs.  What I was getting at here was that the pushdowns for this do not support boolean expressions.  In theory we could parallelize that so that if someone wrote a query that had a series of `OR` conditions, that would be parallelized, but we'll save that for another PR. ;-)

paul-rogers commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -1,346 +1,484 @@
+ */
+public class SimpleHttp {
+  private static final Logger logger = LoggerFactory.getLogger(SimpleHttp.class);
+  private static final Pattern URL_PARAM_REGEX = Pattern.compile("\\{(\\w+)=?(\\w*)\\}");
```
Nit: I think this wants to be `"\\{(\\w+)(?:=(\\w*))?\\}"`

This makes it clear that the second word is allowed only if there is an `=`.

Also, the second word uses `*`. This means that the following is valid: `{foo=}`. This is good, it means that an empty string is a valid default. This works, however, only if the code which checks for a value can tell the difference between no default and a default of a blank string. Can it?

cgivre replied:
Updated regex

paul-rogers commented on contrib/storage-http/src/test/java/org/apache/drill/exec/store/http/TestURLParameters.java:
```diff
@@ -0,0 +1,121 @@
+    HttpUrl githubSingleParam = HttpUrl.parse("https://github.com/orgs/{ORG}/{org}/repos");
+    CaseInsensitiveMap<String> filters = CaseInsensitiveMap.newHashMap();
+    filters.put("org", "apache");
+    filters.put("ORG", "linux");
```
Thanks for the case-sensitivity tests.

paul-rogers commented on contrib/storage-http/src/test/java/org/apache/drill/exec/store/http/TestURLParameters.java:
```diff
@@ -0,0 +1,121 @@
+
+  @Test
+  public void testDefaultParameterExtractor() {
+    HttpUrl pokemonUrl = HttpUrl.parse("https://pokeapi.co/api/{pokemon_name=Misty}");
```
Would be great to test the `{foo=}` case as well to define the expected behavior.

cgivre replied:
@paul-rogers 
I added a unit test for this.  Upon reflecting on it a bit, it would seem to me that the "correct" behavior would be to throw an exception in this case.  Having a blank default rather defeats the purpose of having a default at all, so my thinking here, was either specify a default value or don't.

Does that make sense?   In any event, I think I've addressed all your comments.  Thanks for the review!

paul-rogers reviewed (commented):
@cgivre, thanks for the many improvements. Looking pretty good. A few minor comments.

Regarding the "misuse" of query parameters: I see such "misuse" as a feature, not a bug. Consider. Today I must create a new config for each new URL. But, I'm lazy, so I create a "universal" config: `url="{url}"`. Now I can hit any URL and the magic of Drill's schema-on read will figure it out, assuming the returned data is of the right shape.

The goal of the original comment was to get us to realize, "hey, this might be useful. Lets a) document it and b) test it."

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+    String decodedUrl = SimpleHttp.decodedURL(url);
+    Matcher matcher = URL_PARAM_REGEX.matcher(decodedUrl);
+    return matcher.find();
+  }
```
I moved the `null` check however, there are places where we just need to IF there are URL parameters and then other situations to parse them.  I can't imagine the cost would be that great since the number of params is likely to be very small.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -322,6 +336,99 @@ private void setupCache(Builder builder) {
+    Matcher matcher = URL_PARAM_REGEX.matcher(decodedURL);
+    String param;
+    while (matcher.find()) {
+      param = matcher.group();
```
Fixed

cgivre commented:
> @cgivre, thanks for the many improvements. Looking pretty good. A few minor comments.
> 
> Regarding the "misuse" of query parameters: I see such "misuse" as a feature, not a bug. Consider. Today I must create a new config for each new URL. But, I'm lazy, so I create a "universal" config: `url="{url}"`. Now I can hit any URL and the magic of Drill's schema-on read will figure it out, assuming the returned data is of the right shape.
> 
> The goal of the original comment was to get us to realize, "hey, this might be useful. Lets a) document it and b) test it."

Gotcha!  👍 Makes sense.  I was debating putting some sort of URL validation on the config, but I think I'll leave it alone for the time being if you're ok with that.

cgivre commented:
@paul-rogers 
Hey Paul, 
I think I got everything.  Could you please take a look?  Thanks!
